classes/security/Validation.inc.php

```diff
@@ -257,7 +311,33 @@ function generatePasswordResetHash($userId) {
 			// No such user
 			return false;
 		}
-		return substr(md5($user->getId() . $user->getUsername() . $user->getPassword()), 0, 6);
+		// create hash payload
+		$salt = Config::getVar('security', 'salt');
+		
+		// use last login time to create a basic expiry coefficient
+		$lastLogin = strtotime($user->getDateLastLogin());
+		$time = time();
+		$normalised = ($time - ($time % 3600)) + 7200;
+		$diff = floor(($normalised - $lastLogin) / 3600);
```

Comment:
Good point. Thanks for picking up on that.

I've removed the expiration component for the time being. I was trying to avoid having to change the lib/pkp stuff.
The added salt (and the use of the last login time) still adds an extra layer of security.

A proper expiry could be added with a modification to the following lines which actually use the reset hash:
- https://github.com/pkp/ojs/blob/ojs-stable-2_4_6/classes/user/form/LoginChangePasswordForm.inc.php#L35
- https://github.com/pkp/pkp-lib/blob/13fc909010d53be7957afd07f772fdd783da3fbb/pages/login/PKPLoginHandler.inc.php#L195

implementation like so:
```
// create hash
$expiry = time() + PASSWORD_RESET_SECONDS;
$data   = $user->getUsername() . $user->getPassword() . $user->getDateLastLogin() . $expiry;
$hash   = hash_hmac('sha256', $data, $salt) . ':' . $expiry;

// Example: be987f298ca0f8e7017119b9e76084a666e7181116838e4c968a8fcc02383ba9:1441181037

// ---- //

// check hash
list($hash, $expiry) = explode(':', $hash);

if ((int) $expiry > time()) {
  // not expired
  $data = $user->getUsername() . $user->getPassword() . $user->getDateLastLogin() . $expiry;
  if ($hash === hash_hmac('sha256', $data, $salt)) {
    // hash is valid
  } 
}
```